torch/lib/c10d/reducer.cpp

```diff
@@ -1364,33 +1362,25 @@ void Reducer::finalize_backward() {
   // Wait for asynchronous reduction to complete and unflatten contents.
   for (auto& bucket : buckets_) {
     // See Note [DDP Communication Hook]
-    if (comm_hook_ == nullptr) {
-      TORCH_INTERNAL_ASSERT(
-          bucket.work,
-          "Expected bucket.work not to be null. "
-          "This may indicate that allreduce hooks were not properly installed.");
-      bucket.work->wait();
-    } else {
-      TORCH_INTERNAL_ASSERT(
-          bucket.future_work,
-          "Expected bucket.future_work not to be null. "
-          "This may indicate that communication hook was not properly installed.");
-      bucket.future_work->wait();
-
-      auto future_result =
-          comm_hook_->parseHookResult(bucket.future_work->value());
-
-      for (size_t i = 0; i < future_result.size(); i++) {
-        auto& replica = bucket.replicas[i];
-        if (bucket.expect_sparse_gradient) {
-          replica.contents.copy_(future_result[i]);
-        } else {
-          // Reinitialize only `bucket_views_out` with the future_result by
-          // following the same logic in `initialize_buckets`.
-          populate_bucket_views_out(replica, future_result[i]);
-        }
+    TORCH_INTERNAL_ASSERT(
+        bucket.future_work,
+        "Expected bucket.future_work not to be null. "
+        "This may indicate that communication hook was not properly installed.");
+    bucket.future_work->wait();
```

Comment:
@agolynski Although the future work calls wait() here, it still fails on Gloo backend. As shown by the error message in the test failures:
```
RuntimeError: Process 0 exited with error code 10 and exception:
Traceback (most recent call last):
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_gloo_fork#binary,link-tree/torch/testing/_internal/common_distributed.py", line 509, in run_test
    getattr(self, test_name)()
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_gloo_fork#binary,link-tree/torch/testing/_internal/common_distributed.py", line 393, in wrapper
    fn()
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_gloo_fork#binary,link-tree/torch/testing/_internal/common_distributed.py", line 108, in wrapper
    return func(*args, **kwargs)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_gloo_fork#binary,link-tree/torch/testing/_internal/distributed/distributed_test.py", line 3624, in test_accumulate_gradients_no_sync_grad_is_view
    self._test_accumulate_gradients_no_sync(gradient_as_bucket_view=True)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_gloo_fork#binary,link-tree/torch/testing/_internal/distributed/distributed_test.py", line 3588, in _test_accumulate_gradients_no_sync
    step_model(ddp_model, ddp_input, ddp_target)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_gloo_fork#binary,link-tree/torch/testing/_internal/distributed/distributed_test.py", line 3563, in step_model
    loss.backward()
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_gloo_fork#binary,link-tree/torch/_tensor.py", line 255, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/distributed_gloo_fork#binary,link-tree/torch/autograd/__init__.py", line 147, in backward
    Variable._execution_engine.run_backward(
RuntimeError: Work needs to be completed before calling result(). Should call wait() before result().
```

Could you please take a look?

Reply:
The bug in gloo backend is fixed now.